### PR TITLE
fix nan in LBFGS by adding retries

### PR DESCRIPTION
**Pull request by McHaillet**
This PR adds the following:

* epsilon value to std to prevent division by zero for safety
* some checks to ensure we report NaN values
* removing strong_wolfe line search
* remove nan checks in optimize_iterative; instead raise ValueError in optimize_global in case any of the alignment parameters went to nan after optimization

After a lot of digging I found that the optimization step of LBFGS is actually introducing the nan values in the shifts_x and shifts_y parameters. Still not entirely sure why, and also not entirely sure why this started happening now. The only fix I could find thus far is by turning of 'strong_wolfe' line search mode. I feel uncomfortable because Its unclear to me how much this impact all the results we have generated so far. I will start doing some testing.

### CODE REVIEW

**Comment by McHaillet** [comment]
Okay, it seems stable on 10499, as in: no nans encountered in the parameters, visually tomograms look good. I will now run on SHREC. It might be nice to have an automated end-to-end test on SHREC, to see if we get reliable results after updates.

**Comment by dtegunov** [comment]
From my understanding of BGFS, the line search is quite useful for what we're doing. We could implement smarter handling of NaNs like I did in the anchoring code. E. g. when encountering loss = NaN during BFGS, return a large value instead so it takes a step back, or completely reset the solution to the previous non-NaN result.

**Comment by McHaillet** [comment]
Running without the line-search indeed produced some weird results for some tilt-series in SHREC while other optimized fine. I guess we should go back to something like that. I like the idea best of taking a step back but I dont know how easy that is.

Alternatively, thought we could do a restart option. If NaN's are encountered in the LBFGS closure the alignment could raise a AlignmentNanError. If that is caught, we try to restart up to a number of max_retries . If it's still NaN after that, it just return the original data with a loss of torch.inf (?) .

**Comment by McHaillet** [comment]
I tried some things to influence the optimizer, by returning a penalty or removing state, but this all doesn't help. I guess it's already corrupted, and it's not super easy to tamper with the pytorch optimizer.

> Alternatively, thought we could do a restart option. If NaN's are encountered in the LBFGS closure the alignment could raise a AlignmentNanError. If that is caught, we try to restart up to a number of max_retries . If it's still NaN after that, it just return the original data with a loss of torch.inf (?) .

This is a winner though. I added a while loop with a number of retries and a try..except with custom error. As far as I tested, one retry seems enough, but there is a of course chance that if tails multiple times, so a default of max 3 retries seems sensible to me. In case it still fails, the original alignment with a loss of float('inf') is returned, so that optimize_iterative can effectively compare against it but should always prefer previous alignments.

Running on SHREC now gives the correct results again.

**Comment by McHaillet** [comment]
iter2 (with downsample=2 and anchoring) is at 1.6 +- 0.25 Å
